### ISSUE TITLE
Speed of ddccontrol

### ISSUE THREAD

**Issue by ThomasSeeker** [ddccontrol/ddccontrol#69]
Hello there!

First of all, thank you very much for this wonderful little tool! :-)

I have a question about the speed.
I have version 0.4.4 running, but the speed is quite low (1sec). 
Especially compared to the gddccontrol I tried with version 0.4.2. ... there the reaction time of the monitor when moving a slider was more or less instantaneous.

(BTW, with version 0.4.4 gddccontrol is not working here anymore.)

Is there something I can do to improve the speed? 

```
$ time ddccontrol -r 0x10 -w 70 dev:/dev/i2c-2 -v
ddccontrol version 0.4.4
Copyright 2004-2005 Oleg I. Vdovikin (oleg@cs.msu.su)
Copyright 2004-2006 Nicolas Boichat (nicolas@boichat.ch)
This program comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of this program under the terms of the GNU General Public License.

ddcpci initing...
Starting /usr/bin/ddcpci 1 1225874189 &...
Reading EDID and initializing DDC/CI at bus dev:/dev/i2c-2...
Serial number: 14747457653
Manufactured: Week 12, 2015
EDID version: 1.3
Maximum size: 53 x 30 (cm)
I/O warning : failed to load external entity "/usr/share/ddccontrol-db/monitor/ACR0216.xml"
Document not parsed successfully.
Control colortemp has been discarded by the caps string.
Control hsize has been discarded by the caps string.
Control vsize has been discarded by the caps string.
Control hpincushion has been discarded by the caps string.
Control hpincushionbalance has been discarded by the caps string.
Control vlinearity has been discarded by the caps string.
Control vlinearitybalance has been discarded by the caps string.
Control hmisconvergence has been discarded by the caps string.
Control vmisconvergence has been discarded by the caps string.
Control keybalance has been discarded by the caps string.
Control key has been discarded by the caps string.
Control tdistortionctrl has been discarded by the caps string.
Control tdistortionbalance has been discarded by the caps string.
Control bdistortionctrl has been discarded by the caps string.
Control bdistortionbalance has been discarded by the caps string.
Control tilt has been discarded by the caps string.
Control hmoire has been discarded by the caps string.
Control vmoire has been discarded by the caps string.
Control focus has been discarded by the caps string.
Control settings has been discarded by the caps string.
Control audiospeakervolume has been discarded by the caps string.
Control degauss has been discarded by the caps string.
Control osdorientation has been discarded by the caps string.
Control power has been discarded by the caps string.

EDID readings:
        Plug and Play ID:  [(null)]
        Input type: Digital
=============================== WARNING ===============================
Please update ddccontrol-db, or, if you are already using the latest
version, please send the output of the following command to
ddccontrol-users@lists.sourceforge.net:

LANG= LC_ALL= ddccontrol -p -c -d

Thank you.
=============================== WARNING ===============================

Writing 0x10, 0x46(70)...
Control 0x10: +/70/100   [???]
ddcpci being released...
==>ddcpci is quitting.

real    0m1.022s
user    0m0.012s
sys     0m0.294s
```

**Comment by hamster65** [comment]
I'd like to implement an adaptive brightness control for my screen with this, with an external arduino based sensor. But it is indeed way too slow for smooth regulation. 

This one is way faster:
https://github.com/Informatic/python-ddcci

**Comment by kravemir** [comment]
@ThomasSeeker and @hamster65 sorry for the late response!

To improve the speed of ddccontrol, you can use the latest development version using the `master` branch. See "Installation from sources" in README.

The unreleased version contains implementation of daemon, which gets initialized once, and then can react to new commands quickly. As, each invocation of `ddccontrol` without daemon needs to initialize first, and then performs the command. It should work out of the box, after installing the development version from master branch.

However, I had been doing working on this in my own free time,... and, I've got quite demotivated to continue working on the daemon, and contributing to this project, partially because of comments made in issues #5 and #25 by the user called cheater.

But, to be perfectly honest. My political and philosophical view doesn't align with GNU/GPL philosophy, anymore. This GNU/GPL license has got a viral aspect, which requires to open-source all derivative works,... and, if I was about to put an effort into something, then I would like to put it into SW, which can be beneficial for the whole world, and that includes non-profit and proprietary uses. So, my contributions can be also utilized in proprietary projects, e.g. Acer, Aoc, Phillips, LG, Samsung, etc making their own proprietary tool/GUI for their monitors on top of my work. Now, I'm interested only in contribution to projects with a permissive license.

**Comment by ThomasSeeker** [comment]
Ah, thank you @kravemir for the improved speed! ... I will test it with my next system-upgrade.

And thank you very much for your work. I use ddccontrol regularly and it is just a wonderful little tool! I wouldn't want to miss it anymore! :-)

Concern your conflict with this other guy. I read into it, but I lack the knowledge to decide. I guess he just wants to help. Just an idea: Maybe you can ask in some forum, whether or not the guy has a valid point?

**Comment by kravemir** [comment]
> Ah, thank you @kravemir for the improved speed! ... I will test it with my next system-upgrade.
> 
> And thank you very much for your work. I use ddccontrol regularly and it is just a wonderful little tool! I wouldn't want to miss it anymore! :-)

Appreciated :-) ... though, I have different priorities now, where I put my free time. So, I won't contribute again for some time. The most likely, I would just have rewritten it from a scratch using golang, and with a better design of model and API, and licensed it with non-viral license, e.g. MIT/Apache/BSD/...

> Concern your conflict with this other guy. I read into it, but I lack the knowledge to decide. I guess he just wants to help. Just an idea: Maybe you can ask in some forum, whether or not the guy has a valid point?

Yes, he did have a valid point, but an inappropriate attitude - just criticism, and not contributing any design to secure the daemon, or any code/implementation improving these concerns. And, my reaction wasn't appropriate either, as I was quite attached to the work I've done there in my own free time, and went a bit defensive.

Maybe, @cheater himself can do some real work, and do actually lead and do implementation of some solution, instead of just pointing on shortcomings.

**Comment by cheater** [comment]
sorry @kravemir if you're trying to recruit commits by hate-tagging people more than 2 years after they reported some bugs and requested features (none of which have been touched at all) then that's the wrong strategy to achieve your goal and also I don't even own a laptop or use linux any more

**Comment by kravemir** [comment]
> sorry @kravemir if you're trying to recruit commits by hate-tagging people more than 2 years after they reported some bugs and requested features (none of which have been touched at all) then that's the wrong strategy to achieve your goal and also I don't even own a laptop or use linux any more

@cheater Not at all, I didn't mean this as hate, but as a reflection of my personal motivation and interest into this work, and how can someone's comments affect this. I have mentioned my wrong attitude the same as yours wrong attitude. 

You should think, how you phrase your criticism of others work. When you just criticize someone for their effort to improve opensource project, it can discourage people from caring and wanting to contribute.

And, I have decided to stop being a maintainer. I would love to see someone contributing to this. However. Recruitment, or not,... You seemed to be interested in the project, and I was just wondering, whether your interest was to contribute and help, or just a passive opensource consumerism, or did you then just felt the need to criticize something? So, what was it then? What was your intention with these comments?

I wish I have had taken the criticism differently then, and asked you for design of good solution as a first step in the conversation.

**Comment by cheater** [comment]
I'm not even going to engage this

**Comment by hamster65** [comment]
Thanks! I'll try it out.

**Comment by karanimaan** [comment]
> To improve the speed of ddccontrol, you can use the latest development version using the `master` branch. See "Installation from sources" in README.
> 
> The unreleased version contains implementation of daemon, which gets initialized once, and then can react to new commands quickly. As, each invocation of `ddccontrol` without daemon needs to initialize first, and then performs the command. It should work out of the box, after installing the development version from master branch.

Hi there. Has this been implemented in the latest release? I'm using version 0.6.0, and it's still slow.

**Comment by genodeftest** [comment]
Since the port to Gtk3 (#183), the performance of the GUI has drastically improved. You might want to try again with version 2.0.0 or later.

**Comment by larstobi** [comment]
One possible approach could be to support both architectures simultaneously:

- Keep the existing direct-access CLI mode for portability, scripting, debugging, minimal systems, and non-DBus environments.
- Expand the daemon-based backend for desktop integration, caching, monitor discovery acceleration, and improved multi-client behavior.
- Move all shared DDC/CI, MCCS, probing, and EDID logic into a common library used by both backends.

Example architecture:

    libddccontrol
      ├── direct backend (/dev/i2c-*)
      └── daemon backend (DBus)

    ddccontrol CLI
      ├── direct mode
      ├── daemon mode
      └── auto mode

    gddccontrol GUI
      └── daemon mode preferred

This would allow modern desktop-oriented improvements without forcing all users into a mandatory daemon/DBus architecture.

**Comment by genodeftest** [comment]
Since everything is happening on the same machine, and modern computers are quite fast, I don't think the problem is caused by the daemon-based architecture. I'd rather recommend debugging the bottlenecks and optimizing for performance (in case this is still relevant) than redesigning the architecture.

**Comment by larstobi** [comment]
Du kan skrive noe slikt:

Looking back at the original report from OP, I think the discussion drifted a bit away from the actual problem.

The issue being reported here is not about permissions, security, or access control. OP is trying to implement adaptive brightness control and observes that ddccontrol is too slow for smooth adjustments, while python-ddcci performs significantly better.

Since ddccontrol has now moved away from the legacy PCI and AMD ADL backends and uses i2c-dev, it may be worth revisiting where the remaining latency comes from. My suspicion is that much of the overhead is related to startup, monitor probing, capability parsing, XML database lookups, and other initialization work performed on each invocation, rather than the actual DDC/CI transaction itself. This is probably why the daemon is faster.

If that is the case, the most effective improvements for OP’s use case would likely be:

* reducing or caching monitor detection and capability parsing,
* avoiding repeated initialization for every brightness change,
* providing a lightweight API/library path for applications that perform frequent brightness updates.

In other words, the original “speed” concern may be largely independent of the backend choice and instead be a consequence of the amount of work performed before each DDC/CI command is sent.
